src/java.base/share/classes/jdk/internal/math/MathUtils.java

```diff
@@ -0,0 +1,812 @@
+     * 10<sup><i>k</i></sup> &le; 3/4 &middot; 2<sup>{@code e}</sup>
+     * &lt; 10<sup><i>k</i>+1</sup>.
+     * <p>
+     * The result is correct when
+     * -3_606_689 &le; {@code e} &le; 3_150_619.
+     * Otherwise the result is undefined.
+     *
+     * @param e The exponent of 2, which should meet
+     *          -3_606_689 &le; {@code e} &le; 3_150_619 for safe results.
+     * @return &lfloor;log<sub>10</sub>(3/4 &middot;
+     * 2<sup>{@code e}</sup>)&rfloor;.
+     */
+    static int flog10threeQuartersPow2(int e) {
+        return (int) (e * C_10 + A_10 >> Q_10);
+    }
+
+    /**
+     * Returns the unique integer <i>k</i> such that
+     * 2<sup><i>k</i></sup> &le; 10<sup>{@code e}</sup>
+     * &lt; 2<sup><i>k</i>+1</sup>.
+     * <p>
+     * The result is correct when |{@code e}| &le; 1_838_394.
+     * Otherwise the result is undefined.
+     *
+     * @param e The exponent of 10, which should meet
+     *          |{@code e}| &le; 1_838_394 for safe results.
+     * @return &lfloor;log<sub>2</sub>10<sup>{@code e}</sup>&rfloor;.
+     */
+    static int flog2pow10(int e) {
+        return (int) (e * C_2 >> Q_2);
+    }
+
+    /**
+     * Let 10<sup>-{@code k}</sup> = <i>&beta;</i> 2<sup><i>r</i></sup>,
+     * for the unique pair of integer <i>r</i> and real <i>&beta;</i> meeting
+     * 2<sup>125</sup> &le; <i>&beta;</i> &lt; 2<sup>126</sup>.
+     * Further, let <i>g</i> = &lfloor;<i>&beta;</i>&rfloor; + 1.
+     * Split <i>g</i> into the higher 63 bits <i>g</i><sub>1</sub> and
+     * the lower 63 bits <i>g</i><sub>0</sub>. Thus,
+     * <i>g</i><sub>1</sub> =
+     * &lfloor;<i>g</i> 2<sup>-63</sup>&rfloor;
+     * and
+     * <i>g</i><sub>0</sub> =
+     * <i>g</i> - <i>g</i><sub>1</sub> 2<sup>63</sup>.
+     * <p>
+     * This method returns <i>g</i><sub>1</sub> while
+     * {@link #g0(int)} returns <i>g</i><sub>0</sub>.
+     * <p>
+     * If needed, the exponent <i>r</i> can be computed as
+     * <i>r</i> = {@code flog2pow10(-k)} - 125 (see {@link #flog2pow10(int)}).
+     *
+     * @param k The exponent of 10, which must meet
+     *          {@link #K_MIN} &le; {@code e} &le; {@link #K_MAX}.
+     * @return <i>g</i><sub>1</sub> as described above.
+     */
+    static long g1(int k) {
+        return g[k - K_MIN << 1];
+    }
+
+    /**
+     * Returns <i>g</i><sub>0</sub> as described in
+     * {@link #g1(int)}.
+     *
+     * @param k The exponent of 10, which must meet
+     *          {@link #K_MIN} &le; {@code e} &le; {@link #K_MAX}.
+     * @return <i>g</i><sub>0</sub> as described in
+     * {@link #g1(int)}.
+     */
+    static long g0(int k) {
+        return g[k - K_MIN << 1 | 1];
+    }
+
+    /*
+     * The precomputed values for g1(int) and g0(int).
+     * The first entry must be for an exponent of K_MIN or less.
+     * The last entry must be for an exponent of K_MAX or more.
+     */
+    private static final long[] g = {
+        /* -324 */ 0x4F0C_EDC9_5A71_8DD4L, 0x5B01_E8B0_9AA0_D1B5L,
```

Comment:
Not significant, but might this be clearer instead to comment the array elements on the right?
```
    0x4F0C_EDC9_5A71_8DD4L, 0x5B01_E8B0_9AA0_D1B5L, // -324
```